interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/SpelExpressionSuggesterWithTyper.scala

```diff
@@ -0,0 +1,111 @@
+    val typer = new Typer(commonSupertypeFinder, new KeysDictTyper(dict), strictMethodsChecking, staticMethodInvocationsChecking,
+      typeDefinitions, evaluationContextPreparer = null, methodExecutionForUnknownAllowed, dynamicPropertyAccessAllowed)
+    val spelExpression = expression.expression
+    if (normalizedCaretPosition == 0) {
+      return Nil
+    }
+    val previousChar = spelExpression.substring(normalizedCaretPosition - 1, normalizedCaretPosition)
+    val shouldInsertDummyVariable = (previousChar == "#" || previousChar == ".") && (normalizedCaretPosition == spelExpression.length || !spelExpression.charAt(normalizedCaretPosition).isLetter)
+    val input = if (shouldInsertDummyVariable) {
+      insertDummyVariable(spelExpression, normalizedCaretPosition)
+    } else {
+      spelExpression
+    }
+
+    def filterMapByName[V](map: Map[String, V], name: String): Map[String, V] = {
+      if (shouldInsertDummyVariable) {
+        map
+      } else {
+        map.filter { case (key, _) => key.toLowerCase.contains(name.toLowerCase) }
+      }
+    }
+
+    val rawSpelExpression = parser.parseRaw(input)
+    val ast = rawSpelExpression.getAST
+    val nodeInPosition = findNodeInPosition(ast, normalizedCaretPosition)
+
+    val suggestions = nodeInPosition match {
+      case Some(node) => node match {
+        case v: VariableReference =>
+          val filteredVariables = filterMapByName(variables, v.toStringAST.stripPrefix("#"))
+          filteredVariables.toList.map { case (variable, clazzRef) => ExpressionSuggestion(s"#$variable", clazzRef, fromClass = false) }
+        case p: PropertyOrFieldReference =>
+          val prevNodeInPosition = findPrevNodeInPosition(ast, p.getStartPosition)
+          val typedPrevNode = prevNodeInPosition match {
+            case Some(prevNode) => val spelExpressionWithoutNodeInPosition = input.substring(0, prevNode.getEndPosition) + input.substring(p.getEndPosition)
+              val parsedSpelExpressionWithoutNodeInPosition = parser.parseRaw(spelExpressionWithoutNodeInPosition)
+              val typedExpression = typer.typeExpression(parsedSpelExpressionWithoutNodeInPosition, ValidationContext(variables))
+              typedExpression match {
+                case Validated.Valid(collectedTypingResult) => collectedTypingResult.intermediateResults.get(SpelNodeId(prevNode))
+                case _ => None
+              }
+            case _ => None
+          }
+
+          def filterClassMethods(classDefinition: ClazzDefinition): List[ExpressionSuggestion] = {
+            val methods = filterMapByName(classDefinition.methods, p.getName)
+
+            methods.values.flatten
+              .map(m => ExpressionSuggestion(m.name, m.signatures.head.result, fromClass = false))
+              .toList
+          }
+
+          typedPrevNode.map(_.typingResult.withoutValue).map {
```

Comment:
It won't work correctly with `TypeReference` - we should check `TypingResultWithContext.staticContext` as well. Maybe you could add a support for `TypeReference`s? It would be a power feature!

Reply:
What about this?

Reply:
I've prepared simple test for that:
```
    suggestionsFor("T(java.time.Duration).p") shouldBe List(
      suggestion("parse", Typed[String]),
    )
```
but returned type from Typer is Unknown and staticContext is false.
Added todo, so we can do this in another MR

Reply:
Ok, it works now. :)